Intermittent build failures on MacOS during framework creation

Building OSD on MacOS occasionally fails with errors like the following:

> cd /Users/usdonmac/Projects/usd/sunya/osd_test/build/OpenSubdiv-3_3_3/opensubdiv && /Applications/CMake.app/Contents/bin/cmake -E cmake_symlink_library ../lib/libosdGPU.3.\
3.3.dylib ../lib/libosdGPU.3.3.3.dylib ../lib/libosdGPU.dylib
cmake -E touch_nocreate: failed to update "lib/OpenSubdiv.framework/OpenSubdiv".
make[2]: *** [lib/OpenSubdiv.framework/OpenSubdiv] Error 1
make[1]: *** [opensubdiv/CMakeFiles/osd_static_framework.dir/all] Error 2

Looking through the full log (attached), it appears that the build is trying to write both the static and dynamic libraries to the same file in the framework:

> [100%] Linking CXX shared library ../lib/libosdGPU.dylib
[100%] Linking CXX static library ../lib/OpenSubdiv.framework/OpenSubdiv
[100%] Linking CXX static library ../lib/libosdGPU.a
[100%] Linking CXX shared library ../lib/OpenSubdiv.framework/OpenSubdiv

My guess is that the build may be trying to write to the same file concurrently, leading to the error above.

I've reproduced this on 3.3.0 and the 3.4.1+ release candidates when building with make using -j 12 but not with -j1.

--- a/opensubdiv/CMakeLists.txt
+++ b/opensubdiv/CMakeLists.txt
@@ -287,7 +287,7 @@ if (NOT NO_LIB)
             FRAMEWORK true
             INSTALL_NAME_DIR "@rpath/OpenSubdiv.framework/OpenSubdiv"
             INSTALL_RPATH "@executable_path/Frameworks;@loader_path/Frameworks"
-            OUTPUT_NAME OpenSubdiv
+            OUTPUT_NAME OpenSubdiv_static
             CLEAN_DIRECT_OUTPUT true
         )
 
@@ -301,6 +301,19 @@ if (NOT NO_LIB)
              PUBLIC_HEADER  DESTINATION "${CMAKE_INCDIR_BASE}"
              ARCHIVE DESTINATION "${CMAKE_LIBDIR_BASE}")
 
+        foreach(file ${PUBLIC_HEADER_FILES})
+            add_custom_command(TARGET osd_static_framework POST_BUILD
+                COMMAND ${CMAKE_COMMAND} -E copy
+                "${CMAKE_SOURCE_DIR}/opensubdiv/${file}"
+                "$<TARGET_FILE_DIR:osd_static_framework>/Headers/${file}"
+            )
+        endforeach(file)
+
+        add_custom_command(TARGET osd_static_framework POST_BUILD
+            COMMAND ln -sf
+            "Versions/Current/Headers"
+            "$<TARGET_FILE_DIR:osd_static_framework>/../../Headers"
+        )
 
         #shared framework
         if (BUILD_SHARED_LIBS)

